proposals/NNNN-implement-a-rotate-algorithm.md

```diff
@@ -0,0 +1,210 @@
+            rotate_unguarded(p.first, m, l);
+            return p.first;
+        }
+        f = m;
+        m = p.second;
+        p = swap_ranges(f, m, m, l);
+    }
+    return m;
+}
+```
+
+**Bidirectional indices** are a refinement of forward indices that
+additionally support reverse traversal.
+
+The C++ implementation of the rotate algorithm for the BidirectionalIterator 
+(BidirectionalIndexType in Swift's stdlib) may look like this:
+
+```C++
+template <BidirectionalIterator I>
+I rotate(I f, I m, I l, bidirectional_iterator_tag) {
+    reverse(f, m);
+    reverse(m, l);
+    pair<I, I> p = reverse_until(f, m, l);
+    reverse(p.first, p.second);
+    if (m == p.first) return p.second;
+    return p.first;
+}
+```
+
+**Random access indices** access to any element in constant time (both far and fast).
+
+The C++ implementation of the rotate algorithm for the RandomAccessIterator 
+(RandomAccessIndexType in Swift's stdlib) may look like this:
+
+```C++
+template <RandomAccessIterator I>
+I rotate(I f, I m, I l, std::random_access_iterator_tag) {
+    if (f == m) return l;
+    if (m == l) return f;
+    DifferenceType<I> cycles = gcd(m - f, l - m);
+    rotate_transform<I> rotator(f, m, l);
+    while (cycles-- > 0) rotate_cycle_from(f + cycles, rotator);
+    return rotator.m1;
+}
+```
+
+The complexity of the implementation of these algorithms makes the generic rotate algorithm 
+a perfect candidate for the standard library.
+
+## Proposed solution
+
+The Swift standard library should provide generic implementations of the rotate algorithms 
+for all CollectionTypes.
+
+Standard library will provide implementations of algorithm for all 3 index types:
+- ForwardIndexType
+- BidirectionalIndexType
+- RandomAccessIndexType
+
+Different collection types conforms to different index types, therefore the user will use the 
+most optimal algorithm for his collection type.
+
+## Detailed design
+
+Rotated algorithms, structs and extensions will be implemented in `swift/stdlib/public/core/Rotate.swift`
+
+A precondition of this function is that:
+0 <= first <= middle <= last < count
+
+Extension to CollectionType with generic implementation will be added. 
+Implementation will return rotated sequence and an index of the old first element: 
+
+```Swift
+extension CollectionType {
+    @warn_unused_result
+    public func rotatingFirstFrom(middle: Index) -> (SubSequence, Index) {
+        // This should be handled by slicing and rotating slices.
+        // let sequence1 = self[middle..<endIndex]
+        // let sequence2 = self[startIndex..<middle]
+        // let result = [sequence1, sequence2].flatten()
+        // return (result, calculateIndexOfMidPoint())
+    }
+}
+```
+
+Extensions to CollectionType will be added. These extensions will rotate elements in place:
+
+```Swift
+extension CollectionType where Index : ForwardIndexType {
+    @warn_unused_result
+    public mutating func rotatingFirstFrom(middle: Index) -> Index {
```

Comment:
According to the [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines.html#be-grammatical), this should probably read `rotateFirstFrom()`. Same for the other mutating variants below.